core/src/impl/Kokkos_Core.cpp

```diff
@@ -101,19 +101,28 @@ void initialize_internal(const InitArguments& args) {
   const int skip_device = args.skip_device;
   // if the exact device is not set, but ndevices was given, assign round-robin
   // using on-node MPI rank
-  if (use_gpu < 0 && ndevices >= 0) {
-    auto local_rank_str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK");  // OpenMPI
+  if (use_gpu < 0) {
+    auto const* local_rank_str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK");  // OpenMPI
     if (!local_rank_str)
       local_rank_str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK");  // MVAPICH2
     if (!local_rank_str)
       local_rank_str = std::getenv("SLURM_LOCALID");  // SLURM
-    if (local_rank_str) {
-      auto local_rank = std::atoi(local_rank_str);
-      use_gpu         = local_rank % ndevices;
-    } else {
-      // user only gave us ndevices, but the MPI environment variable wasn't
-      // set. start with GPU 0 at this point
-      use_gpu = 0;
+
+    auto const* ctest_kokkos_device_type = std::getenv("CTEST_KOKKOS_DEVICE_TYPE");  // CTest
+    auto const* ctest_resource_group_count_str = std::getenv("CTEST_RESOURCE_GROUP_COUNT");  // CTest
+    if (ctest_kokkos_device_type && ctest_resource_group_count_str && local_rank_str) {
```

Comment:
Why is it not just predicated on whether `CTEST_KOKKOS_DEVICE_TYPE` is defined or not.

Is that so we are able to define `CTEST_KOKKOS_DEVICE_TYPE` with CMake versions < 3.16?  Why do you think it is better to skip rather than emitting an error?

Reply:
Not just versions < 3.16, but also cases where the resource group system is not actually activated on the CTest side. It is only activated if you pass the `--resource-spec-file` argument to CTest.

Reply:
Do you reckon it is better to silently ignore or to error out in that case?

Reply:
If `--resource-spec-file` is not specified, then that just means the user doesn't want to use it, which is not an error. In that case, it then falls back to detecting based on the rank.

`CTEST_KOKKOS_DEVICE_TYPE` just tells Kokkos "this is the name of the GPU resource type specified in the spec file." If anything, perhaps it should be an error for `CTEST_RESOURCE_GROUP_COUNT` to be specified without `CTEST_KOKKOS_DEVICE_TYPE`.